persistence/provider/boltdb/eventstore.go

```diff
@@ -110,6 +129,66 @@ func (r *eventStoreRepository) NextEventOffset(
 	return next, err
 }
 
+// LoadEventsBySource loads the events produced by the specified source with
+// the given handler key and id.
+//
+// d is the optional parameter, it represents ID of the message that was
+// recorded when the instance was last destroyed.
+func (r *eventStoreRepository) LoadEventsBySource(
+	ctx context.Context,
+	hk, id, d string,
+) (eventstore.Result, error) {
+	var (
+		o   uint64
+		err error
+	)
+
+	if d != "" {
+		r.db.View(
+			ctx,
+			func(tx *bbolt.Tx) {
+				if messageIDs, exists := bboltx.TryBucket(
+					tx,
+					r.appKey,
+					eventStoreBucketKey,
+					eventStoreMesssageIDsBucketKey,
+				); exists {
+					v := messageIDs.Get([]byte(d))
+					if v == nil {
+						err = fmt.Errorf(
+							"message with id %s is not found",
+							d,
+						)
+
+						return
+					}
+
+					o = unmarshalUint64(v)
+				}
+			})
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &eventStoreResult{
+		db:     r.db,
+		appKey: r.appKey,
+		pred: func(i *eventstore.Item) bool {
+			ok := hk == i.Envelope.MetaData.Source.Handler.Key &&
+				id == i.Envelope.MetaData.Source.InstanceId
+
+			if d != "" {
+				return ok && i.Offset > o
```

Comment:
This *shouldn't* be necessary at all, because `r.offset` should already take care of it. But I think this is currently "fixing" the fact that the offset isn't incremented +1 (as per the comment above). That's what it looks like anyway, I haven't tried changing any of it.

Same thing is going on in the in-memory provider.